### DISCUSSION TITLE
Hardware selection for a new project

### DISCUSSION THREAD

**Discussion by smurfix** [micropython/micropython#12567] [Hardware & Peripherals]
I'm designing a couple of nontrivial home automation devices. They're going to run MicroPython because I'm fed up with C/C++ and I know Python way better than I'll ever know Rust even if I started switching languages tomorrow, which I won't, so there. 🤓

Anyway. Given that these things should be (a) reliable and (b) not expensive … which MCU would you choose? Alternately, is there a decent list of supported controllers and their capabilities (RAM and flash size, I/O ports, required periphery like a quartz or external flash …)?

I'm looking for CPUs, not pre-made boards (so RP2040 instead of Pico), except when they're too cheap and/or difficult to solder to bother.

**Comment by andrewleech** [comment]
There isn't any pre-made list as such, other than browsing the [download ](https://micropython.org/download) page and looking at chips used - this won't say much about the stability of them but will at least show features.

It's impossible to give any detailed recomendation without some more requirements though; for home automation I'm going to assuming you want network connectivity, but wifi or ethernet?
It's hard to estimate chip flash/ram/peripheral needs without a bit more detail on the sort of functionality being planned.

In my opintion, stm chips have the best stability and broad feature set implemented here. None of them have built in wifi though, though co-processors are possible. There are some with Ethernet and the nucleo boards with this built in give good examples of their use.

For wifi support I think pico-w is the best most stable example, and depending on your size needs I'd say just design your boards to drop a complete pico-w on them.

Alternatively there's a heap of different esp32* chips that provide cheap wifi, often with a lot of external flash and ram and can be built very small (physically), though the IDF library behind them all is a moving target and often has features not working properly.

**Comment by 2dof** [comment]
You should do some list of functionality of Your automation project, and then just select best platform/MCU.  

From automation/control point of view be aware that if You want implement some  control logic (like PLC or digital control) some of the platform during/after program reset/boot can have Hi level on GPIO (for example in  esp32 some pins go [high on boot](https://espeasy.readthedocs.io/en/latest/Reference/GPIO.html))  
If You plan to do some ADC reading directly from AI and using WI-FI simultaneously then will be problems with noise ( on most IOT platforms). 
If You plan to use CAN protocol - then use platforms with CAN peripherals.

**Comment by davefes** [comment]
After several years of working on ESP32 and a few weeks of Pico W work I would still use bare ESP32 WROOM 32U 4MB modules and either solder to the castellations or mount them on "ESP-32S Breakout Board Converter Module" pcbs.

If you have good access then soldering to the bare modules might be easier that getting the bare ESP32 alignment correct on the breakout boards.  How are you with 28AWG wire?

**Comment by smurfix** [comment]
> for example in esp32 some pins go [high on boot](https://espeasy.readthedocs.io/en/latest/Reference/GPIO.html)

Always fun, that.

> How are you with 28AWG wire?

Could be better, I'm way too old for that *censored*, but when you plan on less than a four/five-figure run it's cheaper to use a module anyway.

However I've seen mixed reviews of ESP32 long term stability, so I'm a bit wary of them. Not every real-world problem can be solved with a watchdog timer.

As for connectivity, that's of the "nice to have for OTA upgrades" category. I'm using a wired bus for the IoT signalling. (This stuff needs to work in my own home and I've got walls a mile thick. Very bad for Wifi.) Which is another problem unfortunately, not many MCUs can do one-wire serial with collision detection; I'm still looking for a solution that doesn't require bit-banging.

**Reply by davefes** [reply]
28AWG PVC-covered wire from China has been one of the best "hardware" changes I have made in the last few years.  However, I grew-up with the smell of solder-flux.

Just my take on the ESP32 ... the added layer(s) FreeRTOS and ESP-IDF just makes things a bit more complicated.  Yes, with both software and system watchdog timers AND extensive error trapping I have generally managed to keep ESP32 systems going for long periods.

I am interested to to see how the Pico W without an OS (?) performs in my systems.

What about LoRa?  433MHz goes through walls better.  I have changed several ESP32 WiFi connected systems to LoRa and now trying u-lora on a Pico.

**Reply by smurfix** [reply]
LoRa needs extra hardware. I'll probably use it on the solar-powered stuff outdoors. Indoor, well, I already have a bus cable to where these things are going to go. Gotta power these things somehow, and batteries are off limits here.

**Comment by sophiedegran** [comment]
Using Lora with external Ebyte E32 card (Tx from 1-2W) for long range point to point work very well on ESP32S3 boards.. (driver SX127X).
Have a look on ESP32N8R8 or ESP32N32R8...

**Comment by peterhinch** [comment]
> I've seen mixed reviews of ESP32 long term stability

Many reported problems may be down to power supply problems. When transmitting, ESP32 chips can draw power spikes of 0.5A. People run them on wall-warts designed for charging phones, then complain that the thing falls over occasionally. I wouldn't take any report seriously unless I knew how the test was conducted. That said, the ESP32-S3 can suffer problems with the USB interface disagreeing with the RF section.

We did a lot of work on the original ESP32 doing long running tests of [mqtt-as](https://github.com/peterhinch/micropython-mqtt/blob/master/mqtt_as/README.md) - there were no hardware issues. And we didn't cheat with a WDT,
